### PR TITLE
[FLINK-7783] Don't always remove checkpoints in ZooKeeperCompletedCheckpointStore#recover()

### PR DESCRIPTION
**Pull request by aljoscha**
This is a bit simpler now than described on the issue: We simply never delete checkpoints except when they are subsumed. We always fail completely when we try to retrieve a checkpoint handle, both in `getLatestCheckpoint()` and `getAllCheckpoints()`. The reason for this are incremental state handles: `getAllCheckpoints()` is used to rebuild the dependency information between the incremental state handles when restoring. If we only return a partial list of state handles from `getAllCheckpoints()` then this logic wouldn't work anymore, that's why we have to fail in this method as soon as one handle is not retrievable.

I could change `getLatestCheckpoint()` to ignore broken handles and return an older one (in fact I have the code for this ready) but it doesn't make sense because `getAllCheckpoints()` is also called when `getLatestCheckpoint()` is called, meaning that restoring will fail anyways.

R: @StefanRRichter AND/OR @tillrohrmann

### CODE REVIEW

**Comment by aljoscha** [comment]
@StefanRRichter I'm not sure whether my concern about incremental state handles is actually accurate.

**Comment by StefanRRichter** [comment]
@aljoscha from the top of my head, I think some concerns are invalid. After being added to to the checkpoint store, the incremental handles should be complete and self contained - all placeholder handles are replaced by the real handles. What is, in fact, build up incrementally is the reference count for each sst-file handle when we re-register on recovery. So if we do not register a handle, it does not account for the reference counting, which might be correct if we assume the handle is "dead". However, we should not call `discardState()` on an unregistered handle because if they are unregistered, they are assumed to still own their handles and will wipe all of them without considering other references in the registry. So for most parts, we can simply rely on the remaining handles building up a reasonable count, except for handles that only still occurred in the broken incremental checkpoints. Those are leaking because we lost any way to properly delete them. But I think this kind of leak always existed in the implementation, because we cannot call dispose if we cannot retrieve the checkpoint.

To sum up: it should work, there is a leak to be aware of, but that problem always existed.

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStoreTest.java [hunk 1]
```diff
@@ -183,11 +179,120 @@ public Void answer(InvocationOnMock invocation) throws Throwable {
...
+		ZooKeeperStateHandleStore<CompletedCheckpoint> zooKeeperStateHandleStoreMock = spy(new ZooKeeperStateHandleStore<>(client, storageHelperMock, Executors.directExecutor()));
+		whenNew(ZooKeeperStateHandleStore.class).withAnyArguments().thenReturn(zooKeeperStateHandleStoreMock);
+		doReturn(checkpointsInZooKeeper).when(zooKeeperStateHandleStoreMock).getAllSortedByNameAndLock();
+
+		final int numCheckpointsToRetain = 1;
+
+		// Mocking for the delete operation on the CuratorFramework client
+		// It assures that the callback is executed synchronously
+
+		final EnsurePath ensurePathMock = mock(EnsurePath.class);
```
Comment (StefanRRichter) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStoreTest.java]: I started to become very critical about excessive mocking in tests. I wonder if we could just have true dummy implementations because most (or all) the mocked objects have clear interfaces. We can then reuse the dummies in all test. A smaller case for this could also be `RetrievableStateHandle` that throws exception or returns an object, instead of doing the repetitive mocking thing.

I had quiet some thoughts on this and even consider giving a lunchtalk about my view on the problem. In a nutshell, a test should verify that a contract works as expected and should not care about implementation details, e.g. how the contract is fulfilled. With mocking, we hardwire the "how". Tests should guard against introducing failures, so one half is showing new failure. But the other half is, that tests should not indicate failures as long as the contract is not violated and the test should be valid as is. 

However, with mocks, test can fail when implementation details diverge from the hardwired expectation even if the contract is not violated. So half of the tests value is lost and the test must constantly be refactored with the code to get it green again, thereby also giving potential to introduce errors in the test that cover errors in the actual change. Repetitive mocking makes things even worse.

Comment (aljoscha) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStoreTest.java]: I agree with that! However, here I just copied the other test and removed the later part. If we want to fix that we have to fix the whole suits of tests we have here.

Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStoreTest.java]: I agree with Stefan, that this test looks really hard to maintain. I think the underlying problem is that `ZooKeeperCompletedCheckpointStore` uses internally the `ZooKeeperStateHandleStore` which is hardwired. I guess it would be better if we had a proper interface which can be passed to the `ZooKeeperCompletedCheckpointStore`. Moreover I think, the current design is a bad separation of concerns wrt how  `ZooKeeperCompletedCheckpointStore` and the `ZooKeeperStateHandleStore` work together. My gut feeling is that `ZooKeeperCompletedCheckpointStore` does not have to know anything about `ZooKeeper` at all. Fixing this, should make the components easier to test as well.

Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStoreTest.java]: I think part of the reason why we have all this testing technological debt is exactly because of copying of sub-optimal tests.

**Comment by StefanRRichter** [comment]
@aljoscha So we want to keep the implementation or be a bit more relaxed about the incremental checkpoints?
